eth/protocols/eth/peer.go

```diff
@@ -323,20 +343,157 @@ func (p *Peer) RequestReceipts(hashes []common.Hash, sink chan *Response) (*Requ
 	p.Log().Debug("Fetching batch of receipts", "count", len(hashes))
 	id := rand.Uint64()
 
+	var req *Request
+	if p.version > ETH69 {
+		req = &Request{
+			id:   id,
+			sink: sink,
+			code: GetReceiptsMsg,
+			want: ReceiptsMsg,
+			data: &GetReceiptsPacket70{
+				RequestId:              id,
+				GetReceiptsRequest:     hashes,
+				FirstBlockReceiptIndex: 0,
+			},
+		}
+		p.receiptBuffer[id] = &partialReceipt{request: hashes}
+	} else {
+		req = &Request{
+			id:   id,
+			sink: sink,
+			code: GetReceiptsMsg,
+			want: ReceiptsMsg,
+			data: &GetReceiptsPacket69{
+				RequestId:          id,
+				GetReceiptsRequest: hashes,
+			},
+		}
+	}
+	if err := p.dispatchRequest(req); err != nil {
+		return nil, err
+	}
+
+	return req, nil
+}
+
+// HandlePartialReceipts re-request partial receipts
+func (p *Peer) requestPartialReceipts(id uint64) error {
+	if _, ok := p.receiptBuffer[id]; !ok {
+		return fmt.Errorf("No partial receipt retreival in progress with id %d", id)
+	}
+
+	lastBlock := len(p.receiptBuffer[id].list) - 1
+	lastReceipt := len(p.receiptBuffer[id].list[lastBlock].items)
+
 	req := &Request{
 		id:   id,
```

Comment:
I think the hacky thing in the implementation is we try to reuse the same request ID for the subsequent requests.

How about this: if we detect that the response is incomplete and subsequent requests are required, we internally maintain the mapping between the original request ID and the following IDs, so that the following responses can be correctly detect and be assembled properly?